Retain ServiceAccount secrets on reapply; CI tweaks

dhaiducek opened the pull request:
The primary goal is to fix the ServiceAccount reconcile that is generating an excessive number of Secrets. As it turns out, this appears to only be relevant for Kubernetes v1.22 and older since newer Kubernetes use tokens instead of secrets.

ref: https://issues.redhat.com/browse/ACM-2049

In the process of working on that issue, I also ran into some CI issues and some other things that I sought to address:
- Implement version files (Because parsing a Makefile isn't very elegant)
- Update workflows to use `release-*` branches 
- Makefile tweaks:
  - Conditionally install `bats`
  - Implement `clean-manifests` and update/replace `sed` with Kustomize patches
  - Set `docker-build` architecture since `image-builder` is only built for `linux/amd64`
  - Add `-v` for E2E test runs
  - Add restoring kustomization to `clean` Make target
  - Shorten output for the SVG base64 in the CSV
- Fix customizing the Docker registry port (If the current registry has a different port, the docker push won't work. This adds some logic to make re-running it work consistently, even if the desired configuration changes.)

yiraeChristineKim commented on Makefile:
```diff
@@ -3,10 +3,10 @@
 # - use the VERSION as arg of the bundle target (e.g make bundle VERSION=0.0.2)
 # - use environment variables to overwrite this value (e.g export VERSION=0.0.2)
-VERSION ?= 3.11.1
+VERSION ?= $(shell cat VERSION)
```
Can we add a default? `:=3.11.1`

yiraeChristineKim replied:
ohohohoho nono it is good I just started review I couldn't see that

yiraeChristineKim replied:
can you write readme `docs/upgrading-gatekeeper.md`

dhaiducek replied:
There's a `release.md` doc that was also updated in this PR.

yiraeChristineKim commented on Makefile:
```diff
@@ -3,10 +3,10 @@
 # Replaces Operator version
 # Set this when when there is a new patch release in the channel.
-REPLACES_VERSION ?= 0.2.6
+REPLACES_VERSION ?= $(shell cat REPLACES_VERSION)
```
This too default

yiraeChristineKim commented on Makefile:
```diff
@@ -95,12 +92,21 @@ all: build
+clean-manifests: IMG = quay.io/gatekeeper/gatekeeper-operator:v$(VERSION)
+clean-manifests: kustomize unpatch-deployment bundle
+	# Reset all kustomization.yaml files
+	git restore **/kustomization.yaml
```
lovely

yiraeChristineKim commented on config/olm-install/kustomization.yaml:
```diff
@@ -1,2 +1,27 @@
+    metadata:
+      name: gatekeeper-operator-sub
+    spec:
+      channel: "3.11"
```
Do we have `sed` for` channel: "3.11"`

dhaiducek replied:
Yes--it's in the `deploy-using-olm` target and was updated to change this `kustomization.yaml` file in this PR.

mprahl replied:
Why not set it to `stable` by default?

dhaiducek replied:
Works for me! I was just taking the current value from the YAML.

yiraeChristineKim commented:
Great! Thanks Dale!

yiraeChristineKim commented:
/hold for Matt

mprahl commented on controllers/merge/merge.go:
```diff
@@ -62,6 +64,19 @@ func retainServiceFields(desiredObj, clusterObj *unstructured.Unstructured) erro
 	return nil
 }
 
+func retainServiceAccountFields(desiredObj, clusterObj *unstructured.Unstructured) error {
```
Could you please add a test for this? A unit test should be fine.

dhaiducek commented on scripts/kind-with-registry.sh:
```diff
@@ -4,20 +4,32 @@ set -o errexit
 # desired cluster name; default is "kind"
```
Good question--we should have something that cleans up both the cluster and the registry--I'll take a look.

dhaiducek commented:
Updates:
- Added unit test for the ServiceAccount merging logic
- Added a `delete-test-cluster` target that deletes the Kind cluster and stops/removes the registry container/image
- Remove the unit tests as a prerequisite for `docker-build`
- Changed `docker-build` to use environment variables (minor thing, but it'll make it easier to revert if needed in the future)

yiraeChristineKim commented:
/unhold

yiraeChristineKim commented:
oh wait 5 commits?

dhaiducek commented:
> oh wait 5 commits?

Yes--each is a separate change. And we probably only want to cherry-pick the bug fix, so separating the commits makes sense from that standpoint also.
